pkg/services/ngalert/api/metrics.go

```diff
@@ -0,0 +1,149 @@
+	notificationLatency prometheus.Histogram
+	notifications       *prometheus.CounterVec
+	notificationsFailed *prometheus.CounterVec
+	requestDuration     *prometheus.HistogramVec
+	silences            *prometheus.GaugeVec
+}
+
+func NewMetrics(r prometheus.Registerer) *Metrics {
+	return &Metrics{
+		alerts: promauto.With(r).NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: "grafana",
+			Subsystem: "alerting",
+			Name:      "alerts",
+			Help:      "How many alerts by state.",
+		}, []string{"state"}),
+		alertsInvalid: promauto.With(r).NewCounter(prometheus.CounterOpts{
+			Namespace: "grafana",
+			Subsystem: "alerting",
+			Name:      "alerts_invalid_total",
+			Help:      "The total number of invalid received alerts.",
+		}),
+		alertsReceived: promauto.With(r).NewCounter(prometheus.CounterOpts{
+			Namespace: "grafana",
+			Subsystem: "alerting",
+			Name:      "alerts_received_total",
+			Help:      "The total number of received alerts.",
+		}),
+		notificationLatency: promauto.With(r).NewHistogram(prometheus.HistogramOpts{
+			Namespace: "grafana",
+			Subsystem: "alerting",
+			Name:      "notification_latency_seconds",
+			Help:      "Histogram of notification deliveries",
+			Buckets:   prometheus.DefBuckets,
+		}),
+		notifications: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "grafana",
+			Subsystem: "alerting",
+			Name:      "notifications_total",
+			Help:      "The total number of attempted notfications by integration.",
+		}, []string{"integration"}),
+		notificationsFailed: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "grafana",
+			Subsystem: "alerting",
+			Name:      "notifications_failed_total",
+			Help:      "The total number of failed notfications by integration.",
+		}, []string{"integration"}),
+		requestDuration: promauto.With(r).NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: "grafana",
+				Subsystem: "alerting",
+				Name:      "request_duration_seconds",
+				Help:      "Histogram of requests to the Alerting API",
+				Buckets:   prometheus.DefBuckets,
+			},
+			[]string{"method", "route", "status_code", "backend"},
+		),
+		silences: promauto.With(r).NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: "grafana",
+			Subsystem: "alerting",
+			Name:      "silences",
+			Help:      "The total number of silences by state.",
+		}, []string{"state"}),
+	}
+}
+
+// Instrument wraps a middleware, instrumenting the request latencies.
+func Instrument(
+	method,
+	path string,
+	action interface{},
+	metrics *Metrics,
+) macaron.Handler {
+	normalizedPath := MakeLabelValue(path)
+
+	return func(c *models.ReqContext) {
+		start := time.Now()
+		var res response.Response
+		val, err := c.Invoke(action)
+		if err == nil && val != nil && len(val) > 0 {
+			res = val[0].Interface().(response.Response)
+		} else {
+			res = routing.ServerError(err)
+		}
+
+		// TODO: We could look up the datasource type via our datasource service
```

Comment:
I don't think we should/need to do this - we don't have access to the datasource directly via the URL IIRC.